doc/sphinx/language/gallina-extensions.rst

```diff
@@ -869,167 +866,137 @@ Module system
-
-    .. cmdv:: Module @ident {* @module_binding} := {+<+ @module_expression}
-
-       is equivalent to an interactive module where each :token:`module_expression` is included.
-
-.. cmd:: Module Type @ident
+.. note::
 
-   This command is used to start an interactive module type :token:`ident`.
+  #. Interactive modules and module types can be nested.
+  #. Interactive modules and module types can't be defined inside of :ref:`sections<section-mechanism>`.
+     Sections can be defined inside of interactive modules and module types.
+  #. Hints and notations (:cmd:`Hint` and :cmd:`Notation` commands) can also appear inside interactive
+     modules and module types. Note that with a module definition like:
 
-   .. cmdv:: Module Type @ident {* @module_binding}
+  ::
 
-      Starts an interactive functor type with parameters given by :token:`module_bindings`.
+     Module N : SIG := M.
 
+  or::
 
-Reserved commands inside an interactive module type:
-~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+    Module N : SIG. … End N.
 
-.. cmd:: Include @module
+  hints and the like valid for ``N`` are not those defined in ``M``
+  (or the module body) but the ones defined in ``SIG``.
 
-   Same as ``Include`` inside a module.
+.. _terminating_module:
 
-.. cmd:: Include {+<+ @module}
+Terminating an interactive module or module type definition
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
 
-   This is a shortcut for the command :n:`Include @module` for each :token:`module`.
+.. todo this size is overkill for such a short section
 
-.. cmd:: @assumption_token Inline @assums
-   :name: Inline
+.. todo: description below applies equally to modules and module types?
 
-   The instance of this assumption will be automatically expanded at functor application, except when
-   this functor application is prefixed by a ``!`` annotation.
+Interactive modules are terminated with the :cmd:`End` command, which
+is also used to terminate :ref:`Sections<section-mechanism>`.
+:n:`End @ident` closes the interactive module or module type :token:`ident`.
+If the module type was given, the command verifies that the content of the module
+matches the module type.  If the module is basic (not a
+functor), its components (constants, inductive types, submodules etc.)
+are now available through the dot notation.
 
-.. cmd:: End @ident
+.. exn:: No such label @ident.
+    :undocumented:
 
-   This command closes the interactive module type :token:`ident`.
+.. exn:: Signature components for label @ident do not match.
+    :undocumented:
 
-   .. exn:: This is not the last opened module type.
-      :undocumented:
+.. exn:: This is not the last opened module.
+    :undocumented:
 
-.. cmd:: Module Type @ident := @module_type
+.. exn:: This is not the last opened module type.
+   :undocumented:
 
-   Defines a module type :token:`ident` equal to :token:`module_type`.
+.. cmd:: Include @module_type_inl {* <+ @module_expr_inl }
```

Comment:
```suggestion
Commands to use inside the definition of an interactive module or module type
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

.. cmd:: Include @module_type_inl {* <+ @module_expr_inl }
```

Without this sub-section title, it would seem like we are still within the "Terminating an interactive module..." sub-section. I also suggest that we swap the order of the two sub-sections.

Note that this is in this section that it should be explained that inside a module type definition `Parameter` et al. do not introduce a global axiom, but declare a constant that should be defined by any module of this type. Cf. the example of "module type" above.

Reply:
"Terminating an interactive module ..." is so short it doesn't deserve to be a section or in the TOC.  Also Include and Declare Module are accepted outside interactive definitions.

I tried to make this a subsubsection or a subsubsubsection but I couldn't figure out what characters to use to underline the title to make it so.  These are not right: https://documentation-style-guide-sphinx.readthedocs.io/en/latest/style-guide.html#headings.  If you know what they are, I'll add that info to doc/sphinx/README.md.

I moved the paragraph before the Note and put the title in bold.  Let me know what you think.

Reply:
I mentioned Parameter as item 4 in the Note.  Let me know if that covers it, I don't really know what you'd like to see where.  If this is not close, maybe you could give me some wording?

Reply:
> "Terminating an interactive module ..." is so short it doesn't deserve to be a section or in the TOC.

Then, that's not the way to do it. The choice of characters for the underline doesn't change the level of the heading. You can choose anything and you just need to be consistent in your file.

>Also Include and Declare Module are accepted outside interactive definitions.

In truth, you are always in an interactive module definition, because files are modules themselves.

Reply:
Good.

> The choice of characters for the underline doesn't change the level of the heading.

That's not what the linked webpage says.  That's confusing.  How does one get subsubheaders?

Reply:
Here's an example from Gallina extensions of a header that doesn't appear in the TOC (and slightly smaller than a subsection header):

```
Implicit Argument Binders
+++++++++++++++++++++++++
```

![image](https://user-images.githubusercontent.com/1253341/77361501-9ed4c200-6d0c-11ea-8cca-ef3a37d50d65.png)

Reply:
It's likely just because there was already many levels of headers above. In short, how to get a subsubheader: you need to put a section inside a section inside a section.

Reply:
Still a mystery to me, but we don't have to figure this out in this PR.